docs/rfc/rfc-011-custom-indexing.md

```diff
@@ -0,0 +1,245 @@
+
+Since we have user reports requesting both properties, we have to choose one or
+the other.  Since the primary job of the consensus engine is to correctly,
+robustly, reliablly, and efficiently replicate application state across the
+network, I believe the correct choice is to favor consensus performance.
+
+Inevitably, the question will arise whether we could do both and toggle between
+them with a flag. Arguably this would be the worst-case scenario, requiring us
+to maintain the complexity of two very-different operational concerns.  If our
+goal is that Tendermint should be as simple, efficient, and trustworthy as
+posible, there is not a strong case for making these options configurable.
+
+### Design Principles
+
+Although there is no unique "best" solution to the issues described above,
+there are some specific principles that a solution should include:
+
+1. **A custom indexer should not require integration into Tendermint core.** A
+   developer or node operator can create, build, deploy, and use a custom
+   indexer with a stock build of the Tendermint consensus node.
+
+2. **Custom indexers cannot stall consensus.** An indexer that is slow or
+   stalls cannot slow down or prevent core consensus from making progress.
+
+   The plugin interface must give node operators control over the tolerances
+   for acceptable indexer performance, and the means to detect when indexers
+   are falling outside those tolerances, but indexer failures should "fail
+   safe" with respect to consensus (even if that means the indexer may miss
+   some data, in sufficiently-extreme circumstances).
+
+3. **Custom indexers control which events they index.** A custom indexer is not
+   limited to only the current transaction and block events, but can observe
+   any event published by the node.
+
+4. **Custom indexing is forward-compatible.** Adding new event types or event
+   data to the consensus node should not require existing custom indexers to be
+   rebuilt or modified, unless they want to take advantage of the new data.
+
+5. **Indexers are responsible for answering queries.** An indexer plugin is not
+   required to support the legacy query filter language, nor to be compatible
+   with the legacy RPC endpoints for accessing them.  Any APIs for clients to
+   query a custom index are the responsibility of the indexer, not the node.
+
+### Informal Design Intent
+
+The design principles described above implicate several components of the
+Tendermint node, beyond just the indexer. In the context of [ADR 075][adr075],
+we are re-working the RPC event subscription API to improve some of the UX
+issues discussed above for RPC clients. It is our expectation that a solution
+for pluggable custom indexing will take advantage of some of the same work.
+
+On that basis, the design approach I am considering for custom indexing looks
+something like this (subject to refinement):
+
+1. A custom indexer runs as a separate process from the node.
```

Comment:
I definitely prefer this method because it eases implementation for anyone wishing to create a custom indexer. One question I think we need to answer before we go down this road is what is the rate of event creation? And, if we are asked to expose even more events in the future, perhaps much more granular ones, should we be worried about the extra IPC overhead that we're incurring here? My general sense is no, but at the moment, I have very little understanding about just how noisy our event production is.

Reply:
The overhead of encoding events is negligible. To understand why, consider an in-process indexer (as it stands today): To write events to storage, it already has to convert them into some kind of storage format, and write them out to a file, table, or other storage target. The cost of doing that is on the same order as the cost of encoding and writing data over a socket to a separate (but local) indexer process.

That said, the rate of event generation does bear on configuration. The profile of event generation now seems to oscillate between steady and spike: Steady during the voting phases, then a spike at commit. The commit spikes are the main determinant of how much data we might need to buffer for an indexer to keep up. Given, however, that the node is _already_ generating all those events in-memory, I think we can argue that the issue is not whether there's enough memory available (if there weren't, we'd already be in trouble) but how long we need to keep it around.

If it takes an indexer 1m to deal with a typical-size block and the block time is around 5m, we should safely be able to buffer (say) 3m of data without fear of getting "double blocked" and blowing up RAM. Those numbers are made-up to illustrate the point.

One concern I have is corner cases like the Osmosis epoch, which I think of as a occasional (rare but expected) _very_ large block. Even a properly-tuned buffer for "normal" blocks could get overrun in that case. We have various options to cope with weird fallback cases like that (ranging from "let it burst" to "drop some data" to "spill temporarily to secondary storage during a replay window"). I think these kinds of cases are rare enough that we don't have to build a castle against them.

Reply:
> If it takes an indexer 1m to deal with a typical-size block and the block time is around 5m, we should safely be able to buffer (say) 3m of data without fear of getting "double blocked" and blowing up RAM. Those numbers are made-up to illustrate the point.

Ah, I think I see the disconnect here. Consumers of an event stream like this one _necessarily must_ decouple receiving events from processing them, if processing is nontrivial and would impact their receive rate. The buffer you're describing here has to exist in the indexer, not the node itself.

Reply:
> Ah, I think I see the disconnect here. Consumers of an event stream like this one _necessarily must_ decouple receiving events from processing them, if processing is nontrivial

Yes, that's right. However, "processing" still includes "receiving". Consider, for the sake of discussion, a "no-op" consumer that pulls objects off a websocket (the current transport for event subscribers) and drops them. The "processing" cost is effectively zero, but the rate at which the consumer can accept them is much lower than the rate at which the producer generates them.

The producer must either slow the rate of production, temporarily cache the results until the consumer catches up, or drop items. For the reasons I mentioned in the other thread, eagerly dropping is obviously the wrong answer. The other two are choices we can make: Currently we slow down.

The only way we can be sure the consumer has parity with the producer is if they run in the same process (and even then, we'd have to play funny games with thread pinning to do it reliably with the Go runtime).

> The buffer you're describing here has to exist in the indexer(s), not the node itself. 

Yes, an indexer might _also_ have to buffer work, to cope with its own latencies. Part of the appeal of having the indexer pull is that it insulates the node from any cost incurred by the indexer—and we want that whether we decide to keep throttling or to use a time window.

Reply:
> Consider, for the sake of discussion, a "no-op" consumer that pulls objects off a websocket (the current transport for event subscribers) and drops them. The "processing" cost is effectively zero, but the rate at which the consumer can accept them is much lower than the rate at which the producer generates them.

I don't understand how this could be true. Can you say more? If a producer can emit events down a websocket at some rate, then sure the consumer should be able to consume them at the same rate? It's just bytes, right?

> For the reasons I mentioned in the other thread, eagerly dropping is obviously the wrong answer . . . The only way we can be sure the consumer has parity with the producer is if they run in the same process (and even then, we'd have to play funny games with thread pinning to do it reliably with the Go runtime).

This distills down to a queueing theory problem. You fundamentally _can't_ ensure consumers have parity with producers in an eventing architecture like this. If the consumer isn't keeping up with the event rate, then the producer _must_ drop events. Otherwise, a bad, down, slow, etc. indexer can essentially break a node.

Reply:
> If a producer can emit events down a websocket at some rate, then sure the consumer should be able to consume them at the same rate? It's just bytes, right?

Yes, if we restrict our focus to just the websocket, then rate of send matches rate of receive. We don't need to do anything special there. That isn't the question this document is meant to resolve.

Within the node, events are not generated sequentially and pushed down a channel, they're generated concurrently (and sometimes in groups) by various pieces of the consensus state machine and the application it's harnessed to. We assign order to them by synchronizing at the point of publication, but then the entire consensus machinery stalls while that happens.

If the sequencing itself were the only delay, it wouldn't really matter: Putting the items in sequence is fast. The problem in the current system is that on the _other_ side of that sequencing step, copies of those events are getting pushed out to websocket consumers (or indexers, or whatever). That is _not_ fast, and right now we wait for it.

So the question at hand is: Given that users don't want to wait, are we willing to swap sides and drop events? And if so, what should be the rate-limiting step? It can't simply be "anything slower than a Go channel" (for the reasons I mentioned).

> You fundamentally _can't_ ensure consumers have parity with producers in an eventing architecture like this. 

Right, that's exactly the point of this RFC. This isn't a theoretical impossibility we're trying to overcome, it's a design question about which direction we want to take to work around it (stall/drop).

Reply:
Ah —

> If we drop events immediately when the subscriber is not able to consume them (e.g., because it’s at the other end of an HTTP request, which is many times slower than the rate of production even for local connections), we will drop (lose) most of those events immediately.

I see what you mean — your producer is generating events in memory, and your consumers are at the other end of a network connection. Then indeed it makes sense to buffer events, up to the e.g. p99 of your burst sizes. Sorry for the noise on this point 👍

Reply:
> Right, that's exactly the point of this RFC. This isn't a theoretical impossibility we're trying to overcome, it's a design question about which direction we want to take to work around it (stall/drop).

So, as an example, it's not uncommon for a process to crash in a way that sockets established to that process remain open, but reads and writes against those sockets block basically indefinitely. (At some point the network stack timeouts come into play and kill things, but that's often O(minutes)). If you block on writes, then this would effectively break consensus progress, I think?

Reply:
> So, as an example, it's not uncommon for a process to crash in a way that sockets established to that process remain open, but reads and writes against those sockets block basically indefinitely. (At some point the network stack timeouts come into play and kill things, but that's often O(minutes)). If you block on writes, then this would effectively break consensus progress, I think?

Yes, that can (and does) happen. We do set timeouts on the connections, so (usually) we'll eventually time out and fail the connections—though it takes a while.

I hypothesize that we'd be better off allowing some messages to be dropped, and swapping to a consumer-pull model (instead of producer-push as now). The key question is what probability of loss we can tolerate—if the answer is "no loss is acceptable" then we will just have to deal with pushback. (We can probably improve on it to some extent, but as you very rightly pointed out, we can't eliminate it without taking some loss)

Reply:
> Then indeed it makes sense to buffer events, up to the e.g. p99 of your burst sizes. Sorry for the noise on this point 

No problem—I'm glad you're raising these points. You're raising questions that the language of the doc should address (and I'll continue making updates as we go)

Reply:
Righto, thanks for the context!